### PR TITLE
Introduce queue in PoP

### PR DESCRIPTION
**Pull request by jhyeom26**
For applying system-wide configuration such as `unbonding_delay`, `RequestQueue` is introduced so we can pop delegation, undelegation and redelegation request requested before specific block time, and lazily apply them.

You can test contract inner unit test with the command below.
```sh
cd contracts/hdac-system/pop
# This is for Linux, adjust --target for your system.
cargo test --features std --target x86_64-unknown-linux-gnu contract_queue
```

### CODE REVIEW

**Review by psy2848048 — COMMENTED** [review]
As there is no account-dependency in `step`,  the account of the caller is not important to implement `step()`.
That means, `step_*()` seems to be called by system account.
But, it needs to be tested by system account call...

**Comment by jhyeom26** [comment]
@psy2848048 I think that goes beyond the scope of this PR. This PR includes just a queue feature, not a `step()` feature. The current implementation just breaks delegation behaviors into the request phase and the execution phase so we can implement `step()` later.

In my opinion, that point of view can be treated in another PR implementing `step()` feature.

**Review by psy2848048 — APPROVED** [review]

**Comment by psy2848048** [comment]
> @psy2848048 I think that goes beyond the scope of this PR. This PR includes just a queue feature, not a `step()` feature. The current implementation just breaks delegation behaviors into the request phase and the execution phase so we can implement `step()` later.
> 
> In my opinion, that point of view can be treated in another PR implementing `step()` feature.

As you need proper rest, could you please any plan of `step()` implementation?
The, I'll work `step_claim()` first with queueing, and `step()` in each step.

In my guess, it would be:

```rust
pub fn step() -> Result<()> {
    // system user: PublicKey([0, 0, 0, ... , 0]) 32 of 0s
    if maybe_system_user != SYSTEM_ACCOUNT {
        runtime::revert(ApiError::NoAccessRights);
    }

    let blocktime = runtime::get_blocktime();

    step_undelegate(blocktime);
    step_redelegate(blocktime);

    distribute(); // will be patched with blocktime
    step_claim();

    Ok(())
}
```

File: execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+};
+
+#[derive(Default, PartialEq)]
+pub struct RequestQueue<T: RequestKey>(pub Vec<RequestQueueEntry<T>>);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct RequestQueueEntry<T: RequestKey> {
+    pub request_key: T,
+    pub amount: U512,
+    pub timestamp: BlockTime,
```
Comment (nowooj) [inline execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:17]: Since we use the `ulb` concept, wouldn't the `block height` fit more than the `block time`?

File: execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        Ok((self.request_key.to_bytes()?.into_iter())
+            .chain(self.amount.to_bytes()?)
+            .chain(self.timestamp.to_bytes()?)
+            .collect())
+    }
+}
+
+impl<T: RequestKey> CLTyped for RequestQueueEntry<T> {
+    fn cl_type() -> CLType {
+        CLType::Any
```
Comment (nowooj) [inline execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:113]: How about using `CLType::Tuple3`? 
If `Any` is used, it cannot be serialized and deserialized by `CLType`. 
And if it is changed to a format other than `Any`,
`ToBytes`, `FromBytes` and `CLTyped for RequestQueue<T>` must also be changed according to the `CLType` format.

File: execution-engine/contracts/hdac-system/pop/src/contract_queue/requests.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        Ok((self.delegator.to_bytes()?.into_iter())
+            .chain(self.src_validator.to_bytes()?)
+            .chain(self.dest_validator.to_bytes()?)
+            .collect())
+    }
+}
+
+impl CLTyped for RedelegateRequestKey {
+    fn cl_type() -> CLType {
+        CLType::Any
```
Comment (nowooj) [inline execution-engine/contracts/hdac-system/pop/src/contract_queue/requests.rs:120]: Also, same to RequestQueueEntry.

File: execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs (hunk 1 above)
Comment (jhyeom26) [reply execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:17]: As far as I know, `ulb` is the concept about the height allowing us to assure it must have been finalized. But the undelegation delay serves as an economic variable or duration for slashing.
So I thought it works in a time concept such as 1 day, 10 days, etc.

I am curious more specifically why do you think the block height is fitter than time.

And I think this needs to be discussed in our team channel.

File: execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs (hunk 2 above)
Comment (jhyeom26) [reply execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:113]: > If Any is used, it cannot be serialized and deserialized by CLType.

I don't understand this point. They are already serialized and deserialized through `ToBytes` and `FromBytes` as `CLType::Any`.

The `CLType`s other than `CLType::Any` provide `ToBytes` and `FromBytes` implementations in the sdk level and they also allow us to implement `CLType` for our own user-defined type as `CLType::Any`.

And `RequestQueueEntry` is a generic struct, not a tuple.
I recommend you to check [this](https://github.com/hdac-io/CasperLabs/blob/94b5be1f53eaf53af05c262e624b867a7deb06c7/execution-engine/types/src/cl_type.rs#L647) out.

Comment (nowooj) [reply execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:113]: It doesn't matter if `RequestKey` is not used outside of execution-engine. If used externally, `CLType::Any` cannot be deserialized.
I think about it now, may not need external access because the required data can be created and thrown at the step function.
If only the key of the queue is used, it may not be necessary, but if possible, keep the serialized format for `CLType`.

Comment (jhyeom26) [reply execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:113]: 1. This is an internal data periodically flushed so it doesn't require deserialization outside of EE.
2. If we choose to use a tuple, then we cannot add associated functions or methods on that type and also we cannot bestow type safety.
3. The reason for implementing CLType in this context is to store the data into EE storage not to interact with a client. I think it is overkill to prohibit defining a struct as much as possible due to the possibility of interaction. I think it is enough to change them or create a new interface when the necessity occurs.

Comment (nowooj) [reply execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:113]: 1. 👌
2. I don't think so. You can use the generic struct and define the function for that struct. I think CLType is just the type for serialized.
3. The types stored in the store are saved by type, and shouldn't they be stored the same? 

There is no problem with the operation, so I followed.

**Review by nowooj — APPROVED** [review]

File: execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs (hunk 2 above)
Comment (jhyeom26) [reply execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:113]: > 2. I don't think so. You can use the generic struct and define the function for that struct. I think CLType is just the type for serialized.

Nope. The `CLTyped` trait is just a trait describing what the type is.
The actual serialization and deserialization are performed by `ToBytes` and `FromBytes` trait.
So even if the `fn cl_type()` is implemented to return `CLType::Tuple3`, they are serialized and deserialized as implemented in `ToBytes` and `FromBytes`. And if we try to deserialize this data outside EE as `CLType::Tuple3`, it will emit an error.

Comment (nowooj) [reply execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:113]: Of course serialized, deserialized uses `ToBytes` and `FromBytes`.
Check the `stored value` section. Decide how to serialize each according to the `CLTyped`. It is said that you can define a type by combining `CLType` according to your struct.

Comment (jhyeom26) [reply execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs:113]: > It is said that you can define a type by combining CLType according to your struct.

It doesn't say we can define a struct that implements `CLTyped` with `CLType::Tuple3`.

I don't know what is the point of mentioning about `engine_shared::stored_value::StoredValue`. This enum is not implementing `CLTyped` nor a fitting example for this case. This is just an abstraction classifying the type of data stored in the ee storage.

We can define a struct consisting of `CLTyped` entries but we cannot store that struct directly into the ee storage.

To write the data into the ee storage, it requires the type of that data should be `CLTyped`. But if you force to use pre-defined CLType(which means not `CLType::Any`), this means we cannot store our own data type directly into the ee storage.

In other words, if we want to store our own struct type into the ee storage, we should implement `CLTyped` with `CLType::Any` for that struct.
